sdk/ml/azure-ai-ml/azure/ai/ml/entities/_compute/_custom_applications.py

```diff
@@ -0,0 +1,226 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+# pylint: disable=protected-access,redefined-builtin
+
+from typing import Dict, List, Optional
+from azure.ai.ml._restclient.v2022_10_01_preview.models import (
+    CustomService,
+    Docker,
+    Endpoint as RestEndpoint,
+    EnvironmentVariable as RestEnvironmentVariable,
+    EnvironmentVariableType as RestEnvironmentVariableType,
+    Image as RestImage,
+    ImageType as RestImageType,
+    Protocol,
+    VolumeDefinition as RestVolumeDefinition,
+    VolumeDefinitionType as RestVolumeDefinitionType,
+)
+from azure.ai.ml.exceptions import ErrorCategory, ErrorTarget, ValidationException
+from azure.ai.ml.constants._compute import CustomApplicationDefaults
+
+
+class ImageSettings:
+    """Specifies an image configuration for a Custom Application.
+
+    :param reference: Image reference URL.
+    :type reference: str
+    """
+
+    def __init__(self, *, reference: str):
+        self.reference = reference
+
+    def _to_rest_object(self) -> RestImage:
+        return RestImage(type=RestImageType.DOCKER, reference=self.reference)
+
+    @classmethod
+    def _from_rest_object(cls, obj: RestImage) -> "ImageSettings":
+        return ImageSettings(reference=obj.reference)
+
+
+class EndpointsSettings:
+    """Specifies an endpoint configuration for a Custom Application.
+
+    :param target: Application port inside the container.
+    :type target: int
+    :param published: Port over which the application is exposed from container.
+    :type published: int
+    """
+
+    def __init__(self, *, target: int, published: int):
+        self.target = target
+        self.published = published
+        self._validate_endpoint_settings()
```

Comment:
nit: should validation be done before assignment?

Reply:
Yep, agreed.